include/envoy/http/filter.h

```diff
@@ -119,6 +119,42 @@ enum class FilterMetadataStatus {
   Continue,
 };
 
+/**
+ * An interface for a per-connection object that is created by a filter and then managed by the
+ * HTTP connection manager. This pushes singleton handling into the HTTP connection manager.
+ */
+
+class MutableHttpConnection;
+
+class PerConnectionObject {
+public:
+  PerConnectionObject(MutableHttpConnection& mutable_connection)
```

Comment:
I would make this a pure interface. Essentially all of the actual data will be carried by the implementation. Same below. Just a pure interface and pure virtual functions to access the underlying data.

Reply:
Thanks for the feedback. If I am understanding this correctly; you would like to only have pure virtual functions in this class; that take `MutableHttpConnection` object as a parameter?

Same with `MutableHttpConnection` class? where there are pure virtual functions taking a `Network::Connection` object which manipulates the actual socket?

This requires the inheriting class to have access to 'Network::Connection' instance right ?

So, is this the intended use case? ... is this the flow you envision?
```
class Filter1PerConnectionObject : public PerConnectionObject {
public:
...
void foo(MutableHttpConnection& mhconn) override; ?
};
```
and
```
class Filter1MutableHttpConnection : public MutableHttpConnection {
public:
...
void bar(Network::Connection& nconn) override; ?
);
```

Please help clarify. I hope to understand your perspective.

Reply:
That's not what I meant exactly. What I mean is at the interface level you can just have a pure virtual interface for the object. The actual constructor can take a mutable http connection and store it if it wants (or not). 

I don't think you need any inheritance on the mutable connection, that can be stored and used as a normal virtual object.

Reply:
Thanks, I am a bit of a slow thinker :) ... I will get back as soon as I visualize it.